### PR TITLE
Run safety ensurer iso mds checkup on migrate events

### PR DESCRIPTION
**Pull request by JeffreyDevloo**

### CODE REVIEW

File: ovs/celery_run.py [hunk 1]
```diff
@@ -192,36 +191,65 @@ def load_ovs_logger(**kwargs):
...
-                    cache.set(key, entry)
-                    loghandler.info('Updated key {0}'.format(key))
-                else:
-                    cache.delete(key)
-                    loghandler.info('Deleted key {0}'.format(key))
+            registrations = persistent.get(key)
+            if not registrations:
+                continue
+            # Filter out all the tasks are are no longer running within celery
+            running_registrations = []
```
Comment (sjovdnbos) [inline ovs/celery_run.py:234]: This improvement in naming <3

File: ovs/celery_run.py [hunk 2]
```diff
@@ -192,36 +191,65 @@ def load_ovs_logger(**kwargs):
...
+            for registration in registrations:
+                task_id = registration.get('task_id')
+                if task_id and task_id in active_task_ids:
+                    running_registrations.append(registration)
+            if running_registrations:
+                if running_registrations == registrations:
+                    # No changes required to be made
+                    continue
+                persistent.assert_value(key, running_registrations, transaction=transaction)
```
Comment (sjovdnbos) [inline ovs/celery_run.py]: You assert the value `running_registrations`, and immediately afterwards set it to this same value?

File: ovs/lib/helpers/decorators.py [hunk 3]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+
+        @wraps(f)
+        def ensure_single_default_inner(self, *args, **kwargs):
+            """
+            Wrapped function
+            :param self: With bind=True, the celery task result itself is passed in
+            :param args: Arguments without default values
+            :param kwargs: Arguments with default values
+            """
+            ensure_single = EnsureSingle(ensure_single_container, self)
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py:197]: This is always done in the inner functions, so why not do it in the general function?

File: ovs/lib/helpers/decorators.py [hunk 4]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+                timeout_message = '{0} - {1} could not be started within timeout of {2}s'.format(self.message, task_log, timeout)
+                raise EnsureSingleTimeoutReached(timeout_message)
+
+            self.unittest_set_state_waiting()
+
+    def _ensure_job_limit(self, kwargs_dict, timeout, task_logs, job_limit=2, registrations_to_skip=0):
+        # type: (dict, float, Tuple[str, str], int, int) -> None
+        """
+        Ensure that the number of jobs don't exceed the limit
+        - Test if the current job limit is not reach
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py]: reach_ed_

File: ovs/lib/helpers/decorators.py [hunk 5]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+        - If job limit is reached: check if the task should be discarded, callbacked or waiting for a job to start
+        :param kwargs_dict: Dict containing all arguments as key word arguments
+        :type kwargs_dict: dict
+        :param timeout: Polling timeout in seconds
+        :type timeout: float
+        :param task_logs: Log identification for the tasks
+        :type task_logs: Tuple[str, str]
+        :param job_limit: Number of jobs to keep running at any given time
+        :type job_limit: int
+        # @todo is this sane? Can't it be done with just the job limit?
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py]: Indeed, both cover the same use as far as i see

File: ovs/lib/helpers/decorators.py [hunk 6]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+                slept += self.poll_sleep_time
+                time.sleep(self.poll_sleep_time)
+        finally:
+            self.run_hook('after_validation')
+
+        successful = True
+        try:
+            if slept:
+                self.logger.info('{0} had to wait {1} seconds before being able to start'.format(task_log_name, slept))
+            if slept >= timeout:
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py]: will now log 
` X had to wait Y` before being able to start`
`Could not start Z`
upon timeout. 
```
if slept:
   slept < timeout:
      log dis
   else:
      log dat
```

File: ovs/lib/helpers/decorators.py [hunk 7]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+                    break
+                self.unittest_set_state_waiting()
+                slept += self.poll_sleep_time
+                time.sleep(self.poll_sleep_time)
+        finally:
+            self.run_hook('after_validation')
+
+        successful = True
+        try:
+            if slept:
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py:836]: Same goes for this construction as previous comment

File: ovs/lib/helpers/decorators.py [hunk 8]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+            hook_func()
+
+    def gather_run_time_info(self):
+        # type: () -> None
+        """
+        Retrieve some information only acquirable when going to execute the code
+        """
+        self.now = '{0}_{1}'.format(int(time.time()), ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(10)))
+        self.thread_name = threading.current_thread().getName()
+        self.unittest_mode = os.environ.get('RUNNING_UNITTESTS') == 'True'
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py:991]: this was offloaded to a function. Why would you not use that one?

File: ovs/lib/helpers/decorators.py [hunk 9]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+        Set the function state to executing
+        Only used for unittesting
+        :param state: State to set
+        :type state: str
+        :param value: Value to add
+        :type value: any
+        :return: None
+        """
+        if self.unittest_mode:
+            self.logger.info('Setting {} to {}'.format(self.thread_name, (state, value)))
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py]: {0} and {1}

File: ovs/lib/helpers/decorators.py [hunk 10]
```diff
@@ -122,380 +172,885 @@ def _ensure_single(task_name, mode, extra_task_names=None, global_timeout=300, c
...
+        """
+        self.now = '{0}_{1}'.format(int(time.time()), ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(10)))
+        self.thread_name = threading.current_thread().getName()
+        self.unittest_mode = os.environ.get('RUNNING_UNITTESTS') == 'True'
+        if self.unittest_mode:
+            self.message = 'Ensure single {0} mode - ID {1} - {2} - {{0}}'.format(self.ensure_single_container.mode, self.now, self.thread_name)
+        else:
+            self.message = 'Ensure single {0} mode - ID {1} - {{0}}'.format(self.ensure_single_container.mode, self.now)
+
+    def unittest_set_state(self, state, value):
```
Comment (sjovdnbos) [inline ovs/lib/helpers/decorators.py:997]: Perhaps set None as default for `value`, or does this have flaws?

**Review by sjovdnbos — CHANGES_REQUESTED** [review]

File: ovs/lib/helpers/decorators.py (hunk 8 above)
Comment (JeffreyDevloo) [reply ovs/lib/helpers/decorators.py:991]: Not in the 2.9 branch

File: ovs/lib/helpers/decorators.py (hunk 10 above)
Comment (JeffreyDevloo) [reply ovs/lib/helpers/decorators.py:997]: No need

File: ovs/lib/helpers/decorators.py (hunk 3 above)
Comment (JeffreyDevloo) [reply ovs/lib/helpers/decorators.py:197]: The `self` part is not defined in the general function
The constructor also fetches some runtime parameters which cannot be fetched on load time

`@decorator` is executed when loading in the module, this does not have any runtime info which the ensure single needs.

File: ovs/celery_run.py (hunk 2 above)
Comment (JeffreyDevloo) [reply ovs/celery_run.py]: Mistake on my end

File: ovs/celery_run.py [hunk 11]
```diff
@@ -217,10 +217,20 @@ def _get_registration_update_transaction():
     for key in persistent.prefix(ENSURE_SINGLE_KEY):
         # Yield task registration keys which are <ensure_single_key>_<task_name>_<ensure_single_mode>
         try:
-            registrations = persistent.get(key)
-            if not registrations:
+            initial_registrations = persistent.get(key)
+            if not initial_registrations:
                 continue
             # Filter out all the tasks are are no longer running within celery
+            # @todo might be better to do in update code?
```
Comment (sjovdnbos) [inline ovs/celery_run.py:224]: Yup, please do this in the update code

**Review by sjovdnbos — APPROVED** [review]

File: ovs/celery_run.py (hunk 11 above)
Comment (JeffreyDevloo) [reply ovs/celery_run.py:224]: Once I backport it, I will

**Pull request merged**